src/glum/_glm.py

```diff
@@ -1873,8 +1871,92 @@ def _expand_categorical_penalties(penalty, X, drop_first):
         elif isinstance(X, np.ndarray):
             X = tm.DenseMatrix(X)
 
+        if not hasattr(self, "feature_names_") or not hasattr(self, "term_names_"):
+            self.feature_names_, self.term_names_ = self.create_feature_and_term_names(
+                X
+            )
+
         return X, y, sample_weight, offset, weights_sum, P1, P2
 
+    def create_feature_and_term_names(
+        self, X: ArrayLike
+    ) -> Tuple[List[str], List[str]]:
+        """
+        Get feature names for the input data.
+
+        Stores them in the ``feature_names_`` and ``column_names_`` attributes.
+
+        Parameters
+        ----------
+        X : ArrayLike
+            Input data.
+
+        Returns
+        -------
+        Tuple[List[str], List[str]]
+            Lists of feature names and term names.
+        """
+        feature_names = []
+        term_names = []
+
+        if isinstance(X, pd.DataFrame):
+            for column, dtype in zip(X.columns, X.dtypes):
+                if pd.api.types.is_categorical_dtype(dtype):
+                    for name in _name_categorical_variables(
+                        dtype.categories,
+                        column,
+                        self.drop_first,
+                        categorical_format=self.categorical_format,
+                    ):
+                        feature_names.append(name)
+                        term_names.append(column)
+                else:
+                    feature_names.append(column)
+                    term_names.append(column)
+
+        elif isinstance(X, tm.StandardizedMatrix):
```

Comment:
We should try to minimize the amount of custom logic depending on the tabmat type. Could you instead create a new method in tabmat to deal with this? I think it would actually be very useful metadata to store in the tabmat object itself.